VACMS-7552 update theme build process

tjheffner opened the pull request:
## Description

Closes #7552. Apologies for so many files changed, it's updating both custom admin themes and some build process files. Most relevant files: `gulpfile.es6.js`, `composer.json`, `.tugboat/config.yml`
- Deletes both theme's css assets directory in favor of compiling them to a .gitignored `dist/` directory in the respective themes. This solves the issue of conflicts created every time styles are updated in a PR due to the minified version conflicting. Conflicts may still arise if editing the same source files, as they should.
- Updated all references in both themes to use the new dist/ directory. Includes `*.libraries.yml` and .scss image paths
- Adds scripts for compiling & watching theme to composer.json, removed CMS-theme specific bits from va:web:install
- Adds call to va:theme:compile to va:start and the tugboat.yml files

## Testing done
- [x] verified styles were compiled for vagovadmin and vagovclaro when rebuilding the site locally
- [x] verified styles for both themes are present on tugboat
- [x] verified watch command successfully sees style changes & recompiles & clears drupal caches

## Screenshots
n/a

## QA steps

testing the build command:
- login as admin
- visit PR environment, check centralized content type for example, see that styles are present in vagovadmin theme
- switch theme to vagovclaro, see that styles are present when changed

If styles aren't present, or look a little funky, the compile command did not run as expected. There should be output in the logs with core .pcss files being processed into .css files. This is a required step for vagovclaro to leverage the css variables.

testing the watch command locally:
- Enable vagovclaro as the admin theme
- run `lando composer va:theme:compile` (or the ddev equivalent). A gulp task should start.
- Edit `docroot/themes/custom/vagovclaro/assets/scss/components/_layout.scss`, change the content-header background color. The gulp task should recompile sass and clear drupal cache
- Refresh your page, see the updated color.

### Definition of Done

- [x] Documentation has been updated, if applicable.
- [ ] Tests have been added if necessary.
- [ ] Automated tests have passed.
- [x] Code Quality Tests have passed.
- [x] Acceptance Criteria in related issue are met.
- [x] Manual Code Review Approved.

### Select Team for PR review

- [x] `Platform CMS team`
- [x] `Sitewide CMS Team`
  - [ ] `⭐️ Content ops`
  - [ ] `⭐️ CMS Experience`
  - [ ] `⭐️ Offices`
  - [ ] `⭐️ Product Support`


### Is this PR blocked by another PR?

- [x] `DO NOT MERGE`

### Does this PR need review from a Product Owner

- [ ] `Needs PO review`

### CMS user-facing annoucement

Is an announcement needed to let editors know of this change?
- [ ] Yes, and it's written in issue ____ and queued for publication.
  - [ ] Merge and ping @ rachel-kauff so she's ready to publish after deployment
- [ ] Yes, but it hasn't yet been written
  - [ ] Don't merge yet -- ping @ rachel-kauff to prompt her to write and queue content
- [x] No announcement is needed for this code change.
  - [x] Merge & carry on unburdened by announcements

tjheffner commented:
This appears to have the styles present on tugboat for both themes... feedback extremely welcomed

ElijahLynn commented:
Tugboat scripts do execute the `composer va:web:install` command so this seems like it should work. 

https://github.com/department-of-veterans-affairs/va.gov-cms/blob/58db48d710eba90950811d34940b09b74df8174b/.tugboat/config.yml#L112-L116

swirtSJW commented on composer.json:
```diff
@@ -601,7 +601,9 @@
             "export NODE_EXTRA_CA_CERTS=/etc/pki/tls/certs/ca-bundle.crt PUPPETEER_SKIP_CHROMIUM_DOWNLOAD=TRUE; cd docroot/core && yarn install",
-            "cd docroot/core && yarn build:css"
+            "cd docroot/core && yarn build:css",
+            "cd docroot/themes/custom/vagovadmin && yarn install && yarn build",
+            "cd docroot/themes/custom/vagovclaro && yarn install && yarn build"
```
I like this in general.  However I don't think we want this to be assigned to web:install  web:install is specific to "web"  which is va.gov local.  I don't think we want to hitch CMS needs (this) to that.

For CMS it should probably have its own command like 'va:compile' and then '"va:start" calls '@va:compile'
do we also need a 'va:watch'?

tjheffner replied:
I've moved these to `va:theme:compile` and created a new `va:theme:watch` command. It watches vagovclaro styles. `va:start` calls the `va:theme:compile` command. I also added references to that command in the tugboat config to ensure it gets compiled there too. 

happy to add it anywhere else it may be needed

swirtSJW reviewed (commented):
THis seems like a good start to getting it working locally, and maybe in tugboat.

We need to make sure this or some other solution works for dev/staging/prod  ANd how we'll know when something goes sideways.

tjheffner commented:
- Moved composer stuff to its own commands (`va:theme:compile` and `va:theme:watch`).
- Added calls for the compile command to `va:start` and update & build in the tugboat config.
- Updated gulp watch to work with docker.
- Verified watch command works locally as well as run through composer or lando. Any scss change triggers recompile + cache clear. No browsersync for hot reload, but changes will show on refresh.
- Updated docs

swirtSJW commented:
@tjheffner  I think next step for this are
- [ ] Get sign-off from platform Platform CMS
   - [ ] Works in Lando
   - [ ] Works in DDEV
   - [ ] Works in tugboat
   - [ ]  Works in BRD (dev, staging, prod)  (may need to get someone to pull this to CMS-test)
   - [ ]  How do we know when it breaks?
- [ ] Communicate the dev method maybe on a topic dive?

ndouglas commented:
I think you'll need this on BRD and likely Tugboat; the `yarn` symlink doesn't exist until `va:web:install` is run but `va:theme:compile` calls `yarn`.

I can commit this later or you can, IDC.

```diff
diff --git a/composer.json b/composer.json
index aa573b4c..65dda38f 100644
--- a/composer.json
+++ b/composer.json
@@ -546,6 +546,7 @@
             "@va:start"
         ],
         "va:theme:compile": [
+            "cd bin && ln -sf ../docroot/libraries/yarn/bin/yarn yarn",
             "export NODE_EXTRA_CA_CERTS=/etc/pki/tls/certs/ca-bundle.crt PUPPETEER_SKIP_CHROMIUM_DOWNLOAD=TRUE; cd docroot/core && yarn install",
             "cd docroot/core && yarn build:css",
             "cd docroot/themes/custom/vagovadmin && yarn install && yarn build",
```

tjheffner commented:
- [x] PR environment confirms approach works on tugboat 
- [x] Confirmed that this approach (and the watch command for vagovclaro) work for local dev, lando, ddev
- [x] Confirmed with @ndouglas that styles compile and are available through BRD, using https://test.dev.cms.va.gov/ #7714
- [x] Confirmed any errors in the compile command are reported & surfaced to slack: https://dsva.slack.com/archives/C37M86Y8G/p1643053894090500 / http://jenkins.vfs.va.gov/job/builds/job/cms-test/323/console [intentionally triggered a false response in the command, alert surfaced]

If this all looks good to you @swirtSJW I think we can drop the 'do not merge' label. I need to coordinate with platform/@ndouglas on the exact timing of the merge with https://github.com/department-of-veterans-affairs/devops/pull/10607

swirtSJW reviewed (approved):
This seems great.  Thanks for taking it on and getting it across the finish line.
I see the checkbox is checked for alerting when it breaks but I don't see a test or reference to the alert.

Where will we know if this fails on BRD?  and will it fail the deploy?
